### PR TITLE
Implement a lowering for type operators in the JVM_IR backend

### PR DESCRIPTION
**Pull request by sfs**
This PR moves some of the logic from ExpressionCodegen/codegenUtils to the IR level, where it can potentially be shared with [JS_IR](https://github.com/JetBrains/kotlin/blob/72f7287ad2897fd97da3f24d256076b3ced3ca75/compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/TypeOperatorLowering.kt) and [native](https://github.com/JetBrains/kotlin-native/blob/4da64036c188e3e29dff00ed4dac071fb4412d34/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/TypeOperatorLowering.kt) in the future.

I've also added two tests for the error messages around not-null assertions. They are currently disabled on JVM_IR, since the IMPLICIT_NOTNULL casts are missing. With these casts, "testErrorMessage" succeeds while "testStaticCallErrorMessage" fails, since the source offsets for a static call don't include the class name.

Finally, I've changed "testReifiedSafeAsWithMutable" to use the result of the safe cast - otherwise the test checks for suboptimal code and fails when using the lowering.

The first 5 commits in this PR are from #2339.

### CODE REVIEW

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt [hunk 1]
```diff
@@ -179,26 +182,44 @@ fun IrBuilderWithScope.irEqeqeq(arg1: IrExpression, arg2: IrExpression) =
...
+        IrStatementOrigin.OROR
+    )
+
+fun IrBuilderWithScope.irAndAnd(arg1: IrExpression, arg2: IrExpression) =
+    irIfThenElse(
+        context.irBuiltIns.booleanType,
+        arg1,
+        arg2,
+        irFalse(),
+        IrStatementOrigin.ANDAND
```
[the remark this answers is not in the document]
Comment (sfs) [reply compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: In this case I want to have short-circuit behavior. Is there any use case where we might want to disable it?

[the remark this answers is not in the document]
Comment (neetopia) [reply compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: My concern is if someone else will use it in future whether they will realize this will be short circuited. But since all ANDAND origin have short circuited behavior, it should be fine.

**Comment by max-kammerer** [comment]
@sfs Please rebase on master

**Comment by sfs** [comment]
@max-kammerer rebased to master

**Comment by sfs** [comment]
Rebased to latest master and moved the `NOT_INSTANCEOF` statement origin to `JvmLoweredDeclarationOrigin`.

**Comment by sfs** [comment]
Rebased on master again.

@dnpetrov Is there anything you'd like me to change with the PR?

**Comment by sfs** [comment]
Rebased on master

**Comment by sfs** [comment]
Rebased on master and incorporated 480313210a510a7d65b9d431b53bccd8f73db0dd and e207c963365af6d25e8ea77564eb2438f2d4e633 (#KT-22275)

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt [hunk 2]
```diff
@@ -40,6 +40,9 @@ inline fun IrBuilderWithScope.irLetS(
     nameHint: String? = null,
     body: (IrValueSymbol) -> IrExpression
 ): IrExpression {
+    if (value is IrGetValue)
```
Comment (udalov) [inline compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: Let's remove `irLet` above to avoid different behavior for a similar-looking function (it's unused both in Kotlin and kotlin-native)

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    endOffset = element.endOffset
+            }
+        })
+        return startOffset to endOffset
+    }
+
+    private fun sourceViewFor(declaration: IrDeclaration) =
+        context.psiSourceManager.getKtFile(declaration.fileParent)!!.viewProvider.contents
+
+    // Since Kotline 1.4 we throw NullPointerExceptions instead of more specialized exception classes.
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: typo in "Kotlin"

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt (hunk 2 above)
Comment (sfs) [reply compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: Done

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt (hunk 3 above)
Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: Fixed

File: compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt [hunk 4]
```diff
@@ -20,26 +20,15 @@ import org.jetbrains.kotlin.utils.addToStdlib.assertedCast
...
-}
-
 inline fun IrBuilderWithScope.irLetS(
     value: IrExpression,
     origin: IrStatementOrigin? = null,
     nameHint: String? = null,
     body: (IrValueSymbol) -> IrExpression
 ): IrExpression {
+    if (value is IrGetValue)
+        return body(value.symbol)
```
Comment (udalov) [inline compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: Looks like this change might not be correct in general, if the `value` is an access to a mutable variable which is changed inside the `body`. (Credit to @SvyatoslavScherbina for noticing.) Maybe we could use this logic in a separate method where it'd explicitly say that mutating the subject inside the body is not allowed (not sure how to name it though)?

Comment (sfs) [reply compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt]: Nice catch! Since this is only an optimization, I've just gotten rid of it for now.

It might be better to have a separate cleanup pass for the IR to get rid of unnecessary temporaries instead.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            IrTypeOperator.IMPLICIT_INTEGER_COERCION -> {
+                irImplicitCast(expression.argument.transformVoid(), expression.typeOperand)
+            }
+
+            IrTypeOperator.CAST ->
+                lowerCast(expression.argument.transformVoid(), expression.typeOperand)
+
+            IrTypeOperator.SAFE_CAST -> {
+                irLetS(expression.argument.transformVoid(), IrStatementOrigin.SAFE_CALL) { valueSymbol ->
+                    irIfThenElse(
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: Everything looks good, except minor impact of bytecode size of reified inline functions with 'as?' : +4 instuctions

Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: Yes, previously we would generate a `reifiedOperationMarker` with a kind argument indicating a safe call. Upon inlining this would then be turned into the code we generate here. If you'd prefer, I could change it back to the old behavior?

Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: Now that I think about it, this actually makes a difference when calling the code from Java.
```kotlin
inline fun <refified T>f(x: Any?) = x as? T
```
Right now, calling `f(null)` from Java fails with an exception. With these changes the code runs without exceptions. So it turns out that this isn't just a size issue, we actually need the old behavior for backwards compatibility.

Comment (udalov) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: @sfs Note that inline functions with reified parameters shouldn't be called from Java, since we generate them with ACC_SYNTHETIC in the current backend (and with ACC_PRIVATE before 1.3.40), and IIRC in JVM IR too. So the issue with the behavior inside the function body is only relevant for access via reflection, which should fail. In any case, I think it'd be nice to keep the same bytecode here as in the current backend

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: Totally agree with @udalov. Please note that ACC_SYNTHETIC flag now is missed in IR

Comment (sfs) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt]: I've implemented the changes in question and added an additional test to check that we generate reified operation markers for safe casts to reified types.

**Comment by max-kammerer** [comment]
@sfs Many thanks!
